osquery/core/init.cpp

```diff
@@ -154,50 +161,97 @@ void Initializer::initDaemon() {
-        // The watcher failed, create a worker.
-        watcher.createWorker();
-        watcher.resetCounters();
-      }
-    } while (watcher.ok());
+  // Add a watcher service thread to start/watch an optional worker and set
+  // of optional extensions in the autoload paths.
+  if (Watcher::countExtensions() > 0 || !FLAGS_disable_watchdog) {
+    Dispatcher::getInstance().addService(
+        std::make_shared<WatcherRunner>(argc_, argv_, !FLAGS_disable_watchdog));
+  }
 
+  // If there are no autoloaded extensions, the watcher service will end,
+  // otherwise it will continue as a background thread and respawn them.
+  // If the watcher is also a worker watchdog it will do nothing but monitor
+  // the extensions and worker process.
+  if (!FLAGS_disable_watchdog) {
+    Dispatcher::joinServices();
     // Executation should never reach this point.
     ::exit(EXIT_FAILURE);
   }
 }
 
-bool Initializer::isWorker() { return (getenv("OSQUERYD_WORKER") != nullptr); }
+void Initializer::initWorker(const std::string& name) {
+  // Set the worker's process name.
+  size_t name_size = strlen(argv_[0]);
+  for (int i = 0; i < argc_; i++) {
+    if (argv_[i] != nullptr) {
+      memset(argv_[i], 0, strlen(argv_[i]));
+    }
+  }
+  strncpy(argv_[0], name.c_str(), name_size);
 
-void Initializer::start() {
-  // Load registry/extension modules before extensions.
-  osquery::loadModules();
+  // Start a watcher watcher thread to exit the process if the watcher exits.
+  Dispatcher::getInstance().addService(
+      std::make_shared<WatcherWatcherRunner>(getppid()));
+}
 
-  // Then set the config/logger plugins, which use a single/active plugin.
-  if (!Registry::setActive("config", FLAGS_config_plugin)) {
-    LOG(ERROR) << "Config plugin not found: " << FLAGS_config_plugin;
-    ::exit(EXIT_CATASTROPHIC);
+void Initializer::initWorkerWatcher(const std::string& name) {
+  if (isWorker()) {
+    initWorker(name);
+  } else {
+    // The watcher will forever monitor and spawn additional workers.
+    initWatcher();
   }
+}
+
+bool Initializer::isWorker() { return (getenv("OSQUERYD_WORKER") != nullptr); }
 
-  if (!Registry::setActive("logger", FLAGS_logger_plugin)) {
-    LOG(ERROR) << "Logger plugin not found: " << FLAGS_logger_plugin;
-    ::exit(EXIT_CATASTROPHIC);
+void Initializer::initConfigLogger() {
+  // Use a delay, meaning the amount of milliseconds waited for extensions.
+  size_t delay = 0;
+  // The timeout is the maximum time in seconds to wait for extensions.
+  size_t timeout = atoi(FLAGS_extensions_timeout.c_str());
+  while (!Registry::setActive("config", FLAGS_config_plugin)) {
+    // If there is at least 1 autoloaded extension, it may broadcast a route
+    // to the active config plugin.
+    if (Watcher::countExtensions() == 0 || delay > timeout * 1000) {
+      LOG(ERROR) << "Config plugin not found: " << FLAGS_config_plugin;
+      ::exit(EXIT_CATASTROPHIC);
+    }
+    ::usleep(kExtensionInitializeMLatency * 1000);
```

Comment:
There are a few "* 1000"s here, we're sleeping for microseconds, stepping with milliseconds, and waiting for a configured number of seconds.